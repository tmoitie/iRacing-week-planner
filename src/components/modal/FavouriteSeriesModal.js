// @flow

import * as React from 'react';
import { useTranslation } from 'react-i18next';
import Modal from './Modal';
import Checkbox from '../Checkbox';
import series from '../../data/season.json';

import styles from '../../styles/main.module.scss';

const groupedSeries = (filteredSeries) => filteredSeries.reduce((grouped, single) => {
  if (single.catid === 1) {
    return {
      ...grouped,
      oval: [
        ...grouped.oval,
        single,
      ],
    };
  }
  if (single.catid === 3) {
    return {
      ...grouped,
      dirtOval: [
        ...grouped.dirtOval,
        single,
      ],
    };
  }
  if (single.catid === 4) {
    return {
      ...grouped,
      rx: [
        ...grouped.rx,
        single,
      ],
    };
  }
  if (single.catid === 5) {
    return {
      ...grouped,
      sportsCar: [
        ...grouped.sportsCar,
        single,
      ],
    };
  }
  return {
    ...grouped,
    formulaCar: [
      ...grouped.formulaCar,
      single,
    ],
  };
}, { oval: [], road: [], dirtOval: [], rx: [], sportsCar: [], formulaCar: [] });

type Props = {
  onClose: () => void,
  isOpen?: boolean,
  favouriteSeries?: Array<number>,
  save: (Array<number>) => void,
};

const defaultProps = {
  isOpen: false,
  favouriteSeries: [],
};

export default function FavouriteSeriesModal({ onClose, isOpen, favouriteSeries, save }: Props) {
  const [searchInput, setSearchInput] = React.useState('');
  const { t } = useTranslation();

  const setCheckboxFavourite = (seriesId, newValue) => {
    const newFavorites = [...favouriteSeries];
    const index = newFavorites.indexOf(seriesId);

    if (index === -1 && newValue) {
      newFavorites.push(seriesId);
    }
    if (index !== -1 && newValue === false) {
      newFavorites.splice(index, 1);
    }
    save(newFavorites);
  };

  const handleChange = (e) => {
    e.preventDefault();
    setSearchInput(e.target.value);
  };

  const getFilteredSeries = () => {
    if (searchInput.length > 0) {
      return series.filter((serie) => serie.seriesname.toUpperCase().includes(searchInput.toUpperCase()));
    }
    return series;
  };

  const renderCheckboxes = (cbSeries) => (
    <div className={styles['col-md-6']} key={cbSeries.seriesid}>
      <Checkbox
        id={`favourite-series-${cbSeries.seriesid}`}
        checked={favouriteSeries.indexOf(cbSeries.seriesid) !== -1}
        onChange={(newValue) => {
          setCheckboxFavourite(cbSeries.seriesid, newValue);
        }}
      >
        {t(cbSeries.seriesname)}
      </Checkbox>
    </div>
  );

  return (
    <Modal
      id="favouriteSeriesModal"
      isOpen={isOpen}
      onClose={onClose}
      title={t('Choose favorite series')}
      doneAction={onClose}
    >
      <div className={styles['form-group']}>
        <input
          type="text"
          className={styles['form-control']}
          placeholder={t('Search series')}
          aria-label="search series"
          onChange={handleChange}
          value={searchInput}
        />
      </div>
      <div className={styles['container-fluid']}>
        {groupedSeries(getFilteredSeries()).oval.length > 0 && (
          <>
            <h3>{t('Oval')}</h3>
            <div className={styles.row}>
              {groupedSeries(getFilteredSeries()).oval.map(renderCheckboxes)}
            </div>
          </>
        )}
        {groupedSeries(getFilteredSeries()).sportsCar.length > 0 && (
          <>
            <h3>{t('Sports Car')}</h3>
            <div className={styles.row}>
              {groupedSeries(getFilteredSeries()).sportsCar.map(renderCheckboxes)}
            </div>
          </>
        )}
        {groupedSeries(getFilteredSeries()).formulaCar.length > 0 && (
          <>
            <h3>{t('Formula Car')}</h3>
            <div className={styles.row}>
              {groupedSeries(getFilteredSeries()).formulaCar.map(renderCheckboxes)}
            </div>
          </>
        )}
        {groupedSeries(getFilteredSeries()).dirtOval.length > 0 && (
          <>
            <h3>{t('Dirt Oval')}</h3>
            <div className={styles.row}>
              {groupedSeries(getFilteredSeries()).dirtOval.map(renderCheckboxes)}
            </div>
          </>
        )}
        {groupedSeries(getFilteredSeries()).rx.length > 0 && (
          <>
            <h3>{t('RX')}</h3>
            <div className={styles.row}>
              {groupedSeries(getFilteredSeries()).rx.map(renderCheckboxes)}
            </div>
          </>
        )}
      </div>
    </Modal>
  );
}

FavouriteSeriesModal.defaultProps = defaultProps;
