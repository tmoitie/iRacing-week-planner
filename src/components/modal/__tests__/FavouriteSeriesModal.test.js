import { describe, test } from '@jest/globals';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FavouriteSeriesModal from '../FavouriteSeriesModal';

describe('components/modal/FavouriteSeriesModal', () => {
  test('renders closed', async () => {
    const onClose = jest.fn(() => {});
    render(
      <FavouriteSeriesModal isOpen={false} onClose={onClose} favouriteSeries={[]} save={() => {}} />,
    );
    expect(screen.queryByTestId('favouriteSeriesModal')).not.toBeTruthy();
  });

  test('renders with checkbox saving', async () => {
    const user = userEvent.setup();
    const onClose = jest.fn(() => {});
    const save = jest.fn(() => {});
    render(
      <FavouriteSeriesModal isOpen onClose={onClose} favouriteSeries={[245]} save={save} />,
    );
    expect(await screen.findByTestId('favouriteSeriesModal')).toMatchSnapshot();

    await user.click(screen.getByLabelText('13th Week iRacing Figure GR8'));
    expect(save).toHaveBeenCalledWith([]);

    await user.click(screen.getByLabelText('IndyCar iRacing Series'));
    expect(save).toHaveBeenCalledWith([245, 374]);

    await user.click(screen.getByText('Close'));
    expect(onClose).toHaveBeenCalled();
  });
});
