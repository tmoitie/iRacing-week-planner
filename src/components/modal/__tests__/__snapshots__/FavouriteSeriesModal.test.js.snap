// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/FavouriteSeriesModal renders with checkbox saving 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="favouriteSeriesModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Choose favorite series
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div>
        <input
          aria-label="search series"
          placeholder="Search series"
          type="text"
          value=""
        />
      </div>
      <div
        class="container-fluid"
      >
        <h3>
          Oval
        </h3>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <div
              class="checkbox"
              data-testid="favourite-series-245"
            >
              <label
                for="favourite-series-245"
              >
                <input
                  checked=""
                  id="favourite-series-245"
                  type="checkbox"
                />
                13th Week iRacing Figure GR8
              </label>
            </div>
          </div>
        </div>
        <h3>
          Sports Car
        </h3>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <div
              class="checkbox"
              data-testid="favourite-series-374"
            >
              <label
                for="favourite-series-374"
              >
                <input
                  id="favourite-series-374"
                  type="checkbox"
                />
                IndyCar iRacing Series
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
